usm_wrappers/salt_wrapper.py

```diff
@@ -0,0 +1,453 @@
+
+
+def _get_mon_id_map(unused_mon_ids, minions):
+    mon_id_map = dict(zip(unused_mon_ids, minions))
+    monitors = {}
+    for id, minion in mon_id_map.iteritems():
+        monitors[id] = {
+            'name': minions[minion].get('monitor_name',
+                                        utils.get_short_hostname(minion)),
+            'address': minions[minion]['public_ip']
+        }
+    return mon_id_map, monitors
+
+
+def _add_ceph_mon_pillar_data(mon_id_map, cluster_name, monitors):
+    pillar_data = {}
+    for id, minion in mon_id_map.iteritems():
+        pillar_data[minion] = {'cluster_name': cluster_name, 'mon_id': id,
+                               'mon_name': monitors[id]['name'],
+                               'public_ip': monitors[id]['address']}
+    return pillar_data
+
+
+def setup_ceph_cluster(cluster_name, fsid, minions):
+    '''
+    :: cluster_name = STRING
+    :: fsid = UUID
+    :: minions = {MINION_ID: {'public_ip': IP_ADDRESS,
+                              'cluster_ip': IP_ADDRESS,
+                              'monitor_name': NAME}, ...}
+    '''
+    public_network, cluster_network = check_minion_networks(minions)
+    mon_id_map, monitors = _get_mon_id_map(_MON_ID_LIST, minions)
+
+    cluster_dir = _CEPH_CLUSTER_CONF_DIR + '/' + cluster_name
+    if not os.path.exists(cluster_dir):
+        os.makedirs(cluster_dir)
+
+    conf_file = cluster_dir + '/' + cluster_name + '.conf'
+    _gen_ceph_cluster_conf(conf_file, cluster_name, fsid, monitors,
+                           public_network)
+
+    _gen_keys(cluster_name, fsid, monitors, cluster_dir)
+
+    pillar_data = _add_ceph_mon_pillar_data(mon_id_map, cluster_name, monitors)
+    pillar = {'usm': pillar_data}
+
+    bootstrapped_minion = None
+    for id, minion in mon_id_map.iteritems():
+        out = run_state(local, minion, 'add_ceph_mon',
+                        kwarg={'pillar':
+                               {'usm': {'mon_bootstrap': True,
+                                        minion: pillar_data[minion]}}})
+        if not out:
+            bootstrapped_minion = minion
+            break
+
+    if not bootstrapped_minion:
+        ## TODO: cleanup created dir/files
+        print "bootstraped_minion is empty"
+        return False
+
+    cluster_key_file = cluster_name + '.keyring'
+    bootstrap_osd_key_file = '/var/lib/ceph/bootstrap-osd/' + cluster_key_file
+    cluster_key_path = cluster_dir + '/' + cluster_key_file
+
+    if not pull_minion_file(local, bootstrapped_minion, bootstrap_osd_key_file,
+                            cluster_key_path):
+        print "pull_minion_file fails"
+        ## mon failed to start
+        return False
+
+    minion_set = set(minions)
+    minion_set.remove(bootstrapped_minion)
+    if minion_set:
+        rv = run_state(local, minion_set, 'add_ceph_mon', expr_form='list',
+                       kwarg={'pillar': pillar})
+        if rv:
+            raise Exception('add_ceph_mon state failed somewhere')
```

Comment:
This is almost useless. Why not define an exception class like BootstrapException and say at least that we failed add_ceph_mon